modules/server/src/main/scala/higherkindness/compendium/http/RootService.scala

```diff
@@ -39,22 +40,28 @@ object RootService {
       case req @ POST -> Root / "protocol" / id =>
         Sync[F].recoverWith(
           for {
-            protocol <- req.as[Protocol]
-            exists   <- CompendiumService[F].existsProtocol(id)
-            _        <- CompendiumService[F].storeProtocol(id, protocol)
-            resp     <- exists.fold(Ok(), Created())
+            protocol   <- req.as[Protocol]
+            protocolId <- validateProtocolId(id)(ProtocolIdentifierError)
+            exists     <- CompendiumService[F].existsProtocol(protocolId)
+            _          <- CompendiumService[F].storeProtocol(protocolId, protocol)
+            resp       <- exists.fold(Ok(), Created())
           } yield resp.putHeaders(Location(req.uri.withPath(s"${req.uri.path}")))
         ) {
           case e: org.apache.avro.SchemaParseException => BadRequest(ErrorResponse(e.getMessage))
           case e: org.http4s.InvalidMessageBodyFailure => BadRequest(ErrorResponse(e.getMessage))
-          case _ => InternalServerError()
+          case idError: ProtocolIdentifierError        => BadRequest(ErrorResponse(idError.message))
+          case _                                       => InternalServerError()
         }
 
       case GET -> Root / "protocol" / id =>
-        for {
-          protocol <- CompendiumService[F].recoverProtocol(id)
-          resp     <- protocol.fold(NotFound())(Ok(_))
-        } yield resp
+        Sync[F].recoverWith(for {
```

Comment:
Instead, you could do:

```scala
(for {
  ...
 } yield resp).recoverWith { ... }
```

Reply:
@pirita do you mind if I change it? I only need to import `applicativeError` syntax. The other endpoint for storing protocols also was written that way (using `Sync[F]`)